### PR TITLE
Feature/cc 0006 automatic calculation of the tile number

### PR DESCRIPTION
**Pull request by jennawan**
test not done yet.

### CODE REVIEW

File: src/components/ProFullCourt/index.tsx [hunk 1]
```diff
@@ -30,6 +34,32 @@ const ProFullCourt = () => {
...
   const [size, setSize] = useState({ width: window.innerWidth, height: window.innerHeight });
 
+  const courtAndTileInfo = {
+    beginPointX: (stageMargin - borderLength) * court.courtRatio,
+    beginPointY: (stageMargin - borderLength) * court.courtRatio,
+    endPointX: (stageMargin + courtAreaXLength + borderLength) * court.courtRatio,
+    endPointY: (stageMargin + courtAreaYLength + borderLength) * court.courtRatio,
+    // TO CHANGE LATER: tile size will be passed in instead of hard coding
+    tileSize: 300 * court.courtRatio,
+  };
```
Comment (tatoMa) [inline src/components/ProFullCourt/index.tsx:44]: Do you think other courts also need this formula for the same calculation? If yes, maybe extract it into a common util file for use on all the other courses.

File: src/components/ProFullCourt/index.tsx [hunk 2]
```diff
@@ -30,6 +34,32 @@ const ProFullCourt = () => {
...
+  const debouncedCalculation = useCallback(
+    debounce(() => {
+      canvas = canvasRef.current as unknown as HTMLCanvasElement;
+      if (canvas) {
+        ctx = canvas.getContext("2d");
+        const tileNumResult = tileNumberCalculator(ctx, courtAndTileInfo);
+        // To Delete later, console for preview only
+        console.log(tileNumResult);
+      }
+    }, 500),
```
Comment (tatoMa) [inline src/components/ProFullCourt/index.tsx:59]: This can be a common utility as well I think.

File: src/components/ProFullCourt/index.tsx [hunk 3]
```diff
@@ -30,6 +34,32 @@ const ProFullCourt = () => {
...
+    []
+  );
+
   useLayoutEffect(() => {
```
Comment (tatoMa) [inline src/components/ProFullCourt/index.tsx:63]: This is a great way to use useLayoutEffect. Nice work!

**Review by tatoMa — COMMENTED** [review]
I kind of get what you are trying to do in the tile number calculator. But my question is that there is only a giant method inside of it. Do you think you also need to extract some of the parts to make it into a few methods that are easier to use and test? (Just my thought)

**Review by EmilyGong111 — APPROVED** [review]
Nice work.
Just remind.
When you extract the logic in ProFullCourt to a common util, make sure if borderLength is undefined would not affect the result.
Because currently other courts does not have border.

File: src/components/ProFullCourt/index.tsx (hunk 1 above)
Comment (jennawan) [reply src/components/ProFullCourt/index.tsx:44]: Thank you for this advice. Other courts would need calculation but with different formulas, as this ProFullCourt has a border that other courts do not have. In a future card, the calculation for other courts (if same), will be extracted to a util file.

File: src/components/ProFullCourt/index.tsx (hunk 2 above)
Comment (jennawan) [reply src/components/ProFullCourt/index.tsx]: Thank you! Done.

**Review by tatoMa — COMMENTED** [review]
Nice work of refactoring the calculator utils.  😀

**Review by Beryl-Zhou — APPROVED** [review]

File: src/utils/tileNumberCalculator.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+// set the color that takes most pixels as the tile color
+// by finding the most repeated element in pixelColorInTile array
+const determineTileColor = (arr: Array<string>) => {
+  const hash: hashObject = {};
+  let maxNum = 0;
+  let mostEle = null;
+  for (let i = 0 as number; i < arr.length; i++) {
+    // ignore white, black and blank color
+    arr[i] === "0,0,0,0" || arr[i] === "0,0,0,255" || arr[i] === "255,255,255,255"
```
Comment (devharris7) [inline src/utils/tileNumberCalculator.ts:34]: ["0,0,0,0", '0,0,0,255', '255,255,255,255'].includes(arr[i])

Comment (devharris7) [reply src/utils/tileNumberCalculator.ts:34]: if ()

File: src/utils/tileNumberCalculator.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+};
+
+export const tileNumberCalculator = (
+  ctx: CanvasRenderingContext2D | null,
+  courtAndTileInfo: IcourtAndTileInfo
+) => {
+  const { beginPointX, beginPointY, endPointX, endPointY, tileSize } = courtAndTileInfo;
+  const rgbaResult: hashObject = {};
+  // scan horizontally
+  for (let x = beginPointX; x < endPointX; x += tileSize) {
```
Comment (devharris7) [inline src/utils/tileNumberCalculator.ts:55]: 双指针？

Comment (devharris7) [reply src/utils/tileNumberCalculator.ts:55]: @chui-mel

**Pull request merged**